### PR TITLE
PLT-587:Added Node repave interval support for all infra cluster

### PR DESCRIPTION
**Pull request by SivaanandM**

### CODE REVIEW

File: spectrocloud/resource_cluster_aws.go [hunk 1]
```diff
@@ -167,6 +167,12 @@ func resourceClusterAws() *schema.Resource {
 							Optional:    true,
 							Description: "Maximum number of nodes in the machine pool. This is used for autoscaling the machine pool.",
 						},
+						"node_repave_interval": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     0,
+							Description: "Minimum number of seconds a node should be Ready, before the next node is selected for repave (Applicable only for worker pools)",
```
Comment (nikchern) [inline spectrocloud/resource_cluster_aws.go:174]: we have to document everywhere that default value is 0.

Comment (nikchern) [inline spectrocloud/resource_cluster_aws.go:174]:
```suggestion
							Description: "Minimum number of seconds node should be Ready, before the next node is selected for repave. Default value is `0`",
```

File: spectrocloud/resource_cluster_azure.go [hunk 2]
```diff
@@ -163,6 +163,12 @@ func resourceClusterAzure() *schema.Resource {
 							Required:    true,
 							Description: "Number of nodes in the machine pool.",
 						},
+						"node_repave_interval": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     0,
+							Description: "Minimum number of seconds a node should be Ready, before the next node is selected for repave (Applicable only for worker pools)",
```
Comment (nikchern) [inline spectrocloud/resource_cluster_azure.go:170]:
```suggestion
							Description: "Minimum number of seconds node should be Ready, before the next node is selected for repave. Default value is `0`",
```

File: spectrocloud/resource_cluster_coxedge.go [hunk 3]
```diff
@@ -197,6 +197,12 @@ func resourceClusterCoxEdge() *schema.Resource {
 							Required:    true,
 							Description: "Number of nodes in the machine pool.",
 						},
+						"node_repave_interval": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     0,
+							Description: "Minimum number of seconds a node should be Ready, before the next node is selected for repave (Applicable only for worker pools)",
```
Comment (nikchern) [inline spectrocloud/resource_cluster_coxedge.go:204]:
```suggestion
							Description: "Minimum number of seconds node should be Ready, before the next node is selected for repave. Default value is `0`",
```

File: spectrocloud/resource_cluster_edge_native.go [hunk 4]
```diff
@@ -171,6 +171,12 @@ func resourceClusterEdgeNative() *schema.Resource {
 							Description:  "Update strategy for the machine pool. Valid values are `RollingUpdateScaleOut` and `RollingUpdateScaleIn`.",
 							ValidateFunc: validation.StringInSlice([]string{"RollingUpdateScaleOut", "RollingUpdateScaleIn"}, false),
 						},
+						"node_repave_interval": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     0,
+							Description: "Minimum number of seconds a node should be Ready, before the next node is selected for repave (Applicable only for worker pools)",
```
Comment (nikchern) [inline spectrocloud/resource_cluster_edge_native.go:178]:
```suggestion
							Description: "Minimum number of seconds node should be Ready, before the next node is selected for repave. Default value is `0`",
```

File: spectrocloud/resource_cluster_edge_vsphere.go [hunk 5]
```diff
@@ -180,6 +180,12 @@ func resourceClusterEdgeVsphere() *schema.Resource {
 							Required:    true,
 							Description: "Number of nodes in the machine pool.",
 						},
+						"node_repave_interval": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     0,
+							Description: "Minimum number of seconds a node should be Ready, before the next node is selected for repave (Applicable only for worker pools)",
```
Comment (nikchern) [inline spectrocloud/resource_cluster_edge_vsphere.go:187]:
```suggestion
							Description: "Minimum number of seconds node should be Ready, before the next node is selected for repave. Default value is `0`",
```

File: spectrocloud/resource_cluster_gcp.go [hunk 6]
```diff
@@ -151,6 +151,12 @@ func resourceClusterGcp() *schema.Resource {
 							Required:    true,
 							Description: "Number of nodes in the machine pool.",
 						},
+						"node_repave_interval": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     0,
+							Description: "Minimum number of seconds a node should be Ready, before the next node is selected for repave (Applicable only for worker pools)",
```
Comment (nikchern) [inline spectrocloud/resource_cluster_gcp.go:158]:
```suggestion
							Description: "Minimum number of seconds node should be Ready, before the next node is selected for repave. Default value is `0`",
```

File: spectrocloud/resource_cluster_libvirt.go [hunk 7]
```diff
@@ -184,6 +184,12 @@ func resourceClusterLibvirt() *schema.Resource {
 							Required:    true,
 							Description: "Number of nodes in the machine pool.",
 						},
+						"node_repave_interval": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     0,
+							Description: "Minimum number of seconds a node should be Ready, before the next node is selected for repave (Applicable only for worker pools)",
```
Comment (nikchern) [inline spectrocloud/resource_cluster_libvirt.go:191]:
```suggestion
							Description: "Minimum number of seconds node should be Ready, before the next node is selected for repave. Default value is `0`",
```

File: spectrocloud/resource_cluster_maas.go [hunk 8]
```diff
@@ -143,6 +143,12 @@ func resourceClusterMaas() *schema.Resource {
 							Required:    true,
 							Description: "Number of nodes in the machine pool.",
 						},
+						"node_repave_interval": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     0,
+							Description: "Minimum number of seconds a node should be Ready, before the next node is selected for repave (Applicable only for worker pools)",
```
Comment (nikchern) [inline spectrocloud/resource_cluster_maas.go:150]:
```suggestion
							Description: "Minimum number of seconds node should be Ready, before the next node is selected for repave. Default value is `0`",
```

File: spectrocloud/resource_cluster_openstack.go [hunk 9]
```diff
@@ -173,6 +173,12 @@ func resourceClusterOpenStack() *schema.Resource {
 							Required:    true,
 							Description: "Number of nodes in the machine pool.",
 						},
+						"node_repave_interval": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     0,
+							Description: "Minimum number of seconds a node should be Ready, before the next node is selected for repave (Applicable only for worker pools)",
```
Comment (nikchern) [inline spectrocloud/resource_cluster_openstack.go:180]:
```suggestion
							Description: "Minimum number of seconds node should be Ready, before the next node is selected for repave. Default value is `0`",
```

File: spectrocloud/resource_cluster_vsphere.go [hunk 10]
```diff
@@ -194,6 +194,12 @@ func resourceClusterVsphere() *schema.Resource {
 							//ForceNew: true,
 							Description: "Whether this machine pool is a control plane and a worker. Defaults to `false`.",
 						},
+						"node_repave_interval": {
+							Type:        schema.TypeInt,
+							Optional:    true,
+							Default:     0,
+							Description: "Minimum number of seconds a node should be Ready, before the next node is selected for repave (Applicable only for worker pools)",
```
Comment (nikchern) [inline spectrocloud/resource_cluster_vsphere.go:201]:
```suggestion
							Description: "Minimum number of seconds node should be Ready, before the next node is selected for repave. Default value is `0`",
```

**Review by nikchern — CHANGES_REQUESTED** [review]
Please address review comments. 
I also suggest to calculate hash on common attributes for machine pool in a separate common function and reuse it.

File: spectrocloud/resource_cluster_aws.go [hunk 11]
```diff
@@ -373,6 +377,15 @@ func flattenMachinePoolConfigsAws(machinePools []*models.V1AwsMachinePoolConfig)
 	return ois
 }
 
+func FlattenControlPlaneAndRepaveInterval(isControlPlane *bool, oi map[string]interface{}) {
+	if isControlPlane != nil {
+		oi["control_plane"] = *isControlPlane
+		if !*isControlPlane {
+			oi["node_repave_interval"] = isControlPlane
```
Comment (nikchern) [inline spectrocloud/resource_cluster_aws.go:384]: should be assigned to node_repave_interval

**Review by nikchern — CHANGES_REQUESTED** [review]

**Review by nikchern — CHANGES_REQUESTED** [review]
Just need to fix odd unit test @SivaanandM otherwise looks great now!

File: spectrocloud/resource_cluster_aws.go [hunk 12]
```diff
@@ -373,6 +377,15 @@ func flattenMachinePoolConfigsAws(machinePools []*models.V1AwsMachinePoolConfig)
...
+		}
```
[the remark this answers is not in the document]
Comment (SivaanandM) [reply spectrocloud/resource_cluster_aws.go]: resolved

File: spectrocloud/resource_cluster_aws.go [hunk 13]
```diff
@@ -373,6 +377,15 @@ func flattenMachinePoolConfigsAws(machinePools []*models.V1AwsMachinePoolConfig)
 	return ois
 }
 
+func FlattenControlPlaneAndRepaveInterval(isControlPlane *bool, oi map[string]interface{}) {
```
[the remark this answers is not in the document]
Comment (SivaanandM) [reply spectrocloud/resource_cluster_aws.go]: resolved

**Comment by SivaanandM** [comment]
> Just need to fix odd unit test @SivaanandM otherwise looks great now!

unit test completed

**Comment by SivaanandM** [comment]
> Please address review comments. I also suggest to calculate hash on common attributes for machine pool in a separate common function and reuse it.

done

**Review by nikchern — APPROVED** [review]

**Pull request merged**